AWS s3 exporter component (#730)
grafana/alloy: docs/sources/reference/components/otelcol.exporter.awss3.md

---

<span class="badge docs-labels__stage docs-labels__item">Experimental</span>

# otelcol.exporter.awss3

{{< docs/shared lookup="stability/experimental.md" source="alloy" version="<ALLOY_VERSION>" >}}

`otelcol.exporter.awss3` accepts telemetry data from other `otelcol` components
and writes them to an AWS S3 Bucket.

> **NOTE**: `otelcol.exporter.awss3` is a wrapper over the upstream
> OpenTelemetry Collector Contrib `awss3` exporter. Bug reports or feature requests will
> be redirected to the upstream repository, if necessary.

Multiple `otelcol.exporter.awss3` components can be specified by giving them
different labels.

## Usage

```alloy
otelcol.exporter.awss3 "LABEL" {
  s3_uploader {
    region = "REGION"
    s3_bucket = "BUCKET_NAME"
    s3_prefix = "PREFIX"
  }
}
```

## Arguments

`otelcol.exporter.awss3` supports the following arguments:

Name       | Type       | Description                                      | Default | Required
-----------|------------|--------------------------------------------------|---------|---------
`encoding` | `string`   | Encoding extension to use to marshal data. Overrides the `marshaler` configuration option if set. | `""`  | no

> `encoding_file_ext` | `string` | file format extension suffix when using the `encoding` configuration option. May be left empty for no suffix to be appended. | `""` | no

Reviewer: ```suggestion
`encoding_file_ext` | `string` | File format extension suffix when using the `encoding` configuration option. It can be left empty if a suffix shouldn't be appended. | `""` | no
```

-----------------------|----------------------|--------------------------------------------------------------------------------------|---------
s3_uploader            | [s3_uploader][]      | Configures the AWS S3 bucket details to send telemetry data to.                      | yes
marshaler              | [marshaler][]        | Marshaler used to produce output data.                                               | no
debug_metrics          | [debug_metrics][]    | Configures the metrics that this component generates to monitor its state.           | no

[s3_uploader]: #s3_uploader-block
[marshaler]: #marshaler-block
[debug_metrics]: #debug_metrics-block

### s3_uploader block

The `s3_uploader` block configures the AWS S3 bucket details used by the component.

The following arguments are supported:

Name                  | Type                       | Description                                                                      | Default      | Required
----------------------|----------------------------|----------------------------------------------------------------------------------|--------------|---------

> `region`              | `string`                   | AWS region.                                                                      | `"us-east-1"`| no
> `s3_bucket`           | `string`                   | S3 bucket                                                                        |              | yes
> `s3_prefix`           | `string`                   | Prefix for the S3 key (root directory inside bucket).                            |              | yes
> `s3_partition`        | `string`                   | Time granularity of S3 key: hour or minute                                       | `"minute"`   | no
> `role_arn`            | `string`                   | The Role ARN to be assumed                                                       |              | no
> `file_prefix`         | `string`                   | File prefix defined by user                                                      |              | no
> `endpoint`            | `string`                   | Overrides the endpoint used by the exporter instead of constructing it from `region` and `s3_bucket` |      | no
> `s3_force_path_style` | `boolean`                  | [Set this to `true` to force the request to use path-style addressing](http://docs.aws.amazon.com/AmazonS3/latest/dev/VirtualHosting.html) | `false`             | no
> `disable_ssl`         | `boolean`                  | Set this to `true` to disable SSL when sending requests           |              | `false`
> `compression`         | `string`                   | should the file be compressed                                                    | `none`      | no

Reviewer: ```suggestion
`region`              | `string`                   | The AWS region.                                                                      | `"us-east-1"`| no
`s3_bucket`           | `string`                   | The S3 bucket.                                                                        |              | yes
`s3_prefix`           | `string`                   | Prefix for the S3 key (root directory inside the bucket).                            |              | yes
`s3_partition`        | `string`                   | Time granularity of S3 key: hour or minute.                                       | `"minute"`   | no
`role_arn`            | `string`                   | The Role ARN to be assumed.                                                       |              | no
`file_prefix`         | `string`                   | The file prefix defined by the user.                                                      |              | no
`endpoint`            | `string`                   | Overrides the endpoint used by the exporter instead of constructing it from `region` and `s3_bucket`. |      | no
`s3_force_path_style` | `boolean`                  |  Set this to `true` to force the request to use [path-style requests](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html#path-style-access) | `false`             | no
`disable_ssl`         | `boolean`                  |  Set this to `true` to disable SSL when sending requests           |              | `false`
`compression`         | `string`                   | should the file be compressed                                                    | `none`      | no

[path-style requests]: https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html#path-style-access
```

Added a bunch of missing periods/full stops.

I've updated the external doc reference URL for path style addressing and the link text - is this still correct?

The description for `compression` doesn't look complete to me. It implies a boolean by the description, but it's a string. What do you put in the string? A compression algo? (zip, rar, tgz, etc?)

Author: I'll get the description cleaned up for `compression`. Just pulled from the OTEL docs. But that's a good call out.

### marshaler block

Marshaler determines the format of data sent to AWS S3. Currently, the following marshalers are implemented:

- `otlp_json` (default): the [OpenTelemetry Protocol format](https://github.com/open-telemetry/opentelemetry-proto), represented as json.
- `otlp_proto`: the [OpenTelemetry Protocol format](https://github.com/open-telemetry/opentelemetry-proto), represented as Protocol Buffers. A single protobuf message is written into each object.
- `sumo_ic`: the [Sumo Logic Installed Collector Archive format](https://help.sumologic.com/docs/manage/data-archiving/archive/).
  **This format is supported only for logs.**
- `body`: export the log body as string.
  **This format is supported only for logs.**

The following arguments are supported:

Name                    | Type       | Description                                                                                | Default | Required
------------------------|------------|--------------------------------------------------------------------------------------------|---------|---------

> `type`                  | `string`   | Marshaler used to produce output data                                                      | `"otlp_json"`   | no

Reviewer: ```suggestion
`type`                  | `string`   | Marshaler used to produce output data.                                                      | `"otlp_json"`   | no
```

### debug_metrics block

{{< docs/shared lookup="reference/components/otelcol-debug-metrics-block.md" source="alloy" version="<ALLOY_VERSION>" >}}

### Encoding

> Encoding overrides marshaler if present and sets to use an encoding extension defined in the collector configuration.
> 
> See [https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/encoding](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/encoding).

Reviewer: ```suggestion
Encoding overrides the marshaler if it's present and sets it to use the encoding extension defined in the collector configuration.

Refer to the Open Telemetry [encoding extensions] documentation for more information.

[encoding]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/encoding
```

### Compression

- `none` (default): No compression will be applied

> - `gzip`: Files will be compressed with gzip. **This does not support `sumo_ic`marshaler.**

Reviewer: ```suggestion
- `gzip`: Files will be compressed with gzip. **This doesn't support `sumo_ic`marshaler.**
```

## Exported fields

The following fields are exported and can be referenced by other components:

Name    | Type               | Description
--------|--------------------|-----------------------------------------------------------------
`input` | `otelcol.Consumer` | A value that other components can use to send telemetry data to.

`input` accepts `otelcol.Consumer` data for any telemetry signal (metrics,
logs, or traces).

## Component health

`otelcol.exporter.awss3` is only reported as unhealthy if given an invalid
configuration.

## Debug information

> `otelcol.exporter.awss3` does not expose any component-specific debug

Reviewer: ```suggestion
`otelcol.exporter.awss3` doesn't expose any component-specific debug
```

information.

## Debug metrics

* `exporter_sent_spans_ratio_total` (counter): Number of spans successfully sent to destination.
* `exporter_send_failed_spans_ratio_total` (counter): Number of spans in failed attempts to send to destination.

> * `exporter_queue_capacity_ratio` (gauge): Fixed capacity of the retry queue (in batches)
> * `exporter_queue_size_ratio` (gauge): Current size of the retry queue (in batches)

Reviewer: ```suggestion
* `exporter_queue_capacity_ratio` (gauge): Fixed capacity of the retry queue (in batches).
* `exporter_queue_size_ratio` (gauge): Current size of the retry queue (in batches).
```
